src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,171 @@
 package core.basesyntax;
-
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static class Node<T> {
+        T item;
+        Node<T> next;
+        Node<T> prev;
+
+        public Node(T item, Node<T> next, Node<T> prev) {
+            this.item = item;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+        head = null;
+        tail = null;
+        size = 0;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            if (current == null) {
+                throw new NullPointerException();
+            }
+            current = current.next;
+        }
+        return current;
+    }
+
+
+    private int findIndexByValue(T value) {
+        Node<T> current = head;
+        int index = 0;
+        while (current != null) {
+            if (Objects.equals(value, current.item)) {
+                return index;
+            }
+            current = current.next;
+            index++;
+        }
+        return index;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node.prev == null) {
+            head = node.next;
+        } else {
+            node.prev.next = node.next;
+            if (node.next != null) {
+                node.next.prev = node.prev;
+            } else {
+                tail = node.prev;
+            }
+        }
+        size--;
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value, null, null);
+        if (size == 0) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> newNode = new Node<>(value, null, null);
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size + "Index Less than 0 or Index more than size");
```

Comment:
The error message 'Index Less than 0 or Index more than size' is somewhat misleading because `index == size` is a valid case for adding an element at the end of the list.